fog/block_provider/src/mobilecoind.rs

```diff
@@ -0,0 +1,200 @@
+};
+use mc_transaction_core::tx::{TxOut, TxOutMembershipProof};
+use mc_util_grpc::ConnectionUriGrpcioChannel;
+use mc_watcher::watcher_db::{
+    POLL_BLOCK_TIMESTAMP_ERROR_RETRY_FREQUENCY, POLL_BLOCK_TIMESTAMP_POLLING_FREQUENCY,
+};
+use std::{
+    sync::Arc,
+    thread::sleep,
+    time::{Duration, Instant},
+};
+
+#[derive(Clone)]
+pub struct MobilecoindBlockProvider {
+    client: MobilecoindApiClient,
+    logger: Logger,
+}
+
+impl MobilecoindBlockProvider {
+    pub fn new(mobilecoind_uri: &MobilecoindUri, logger: &Logger) -> Box<Self> {
+        let env = Arc::new(EnvBuilder::new().name_prefix("Mobilecoind-GRPC").build());
+        let ch = ChannelBuilder::new(env)
+            .max_receive_message_len(std::i32::MAX)
+            .max_send_message_len(std::i32::MAX)
+            .connect_to_uri(mobilecoind_uri, logger);
+
+        let client = MobilecoindApiClient::new(ch);
+
+        Box::new(Self {
+            client,
+            logger: logger.clone(),
+        })
+    }
+}
+
+impl BlockProvider for MobilecoindBlockProvider {
+    fn num_blocks(&self) -> Result<u64, Error> {
+        let response = self.client.get_ledger_info(&Default::default())?;
+        Ok(response.block_count)
+    }
+
+    /// Get the latest block in the ledger.
+    fn get_latest_block(&self) -> Result<Block, Error> {
+        let response = self.client.get_latest_block(&Default::default())?;
+        Ok(response.get_block().try_into()?)
+    }
+
+    /// Get block data of multiple blocks by block number, and in addition get
+    /// information about the latest block. Also include block timestamp for
+    /// each block, if available.
+    fn get_blocks_data(&self, block_indices: &[BlockIndex]) -> Result<BlocksDataResponse, Error> {
+        let request = GetBlocksDataRequest {
+            blocks: block_indices.to_vec(),
+            ..Default::default()
+        };
+        let response = self.client.get_blocks_data(&request)?;
+
+        let results = response
+            .results
+            .iter()
+            .map(|result| {
+                if !result.found {
+                    return Ok(None);
+                }
+
+                Ok(Some(BlockDataWithTimestamp {
+                    block_data: BlockData::try_from(result.get_block_data())?,
+                    block_timestamp: result.timestamp,
+                    block_timestamp_result_code: (&result.timestamp_result_code).try_into()?,
+                }))
+            })
+            .collect::<Result<Vec<_>, Error>>()?;
+
+        let latest_block = Block::try_from(response.get_latest_block())?;
+
+        Ok(BlocksDataResponse {
+            results,
+            latest_block,
+        })
+    }
+
+    /// Poll indefinitely for a watcher timestamp, logging warnings if we wait
+    /// for more than watcher_timeout.
+    fn poll_block_timestamp(&self, block_index: BlockIndex, watcher_timeout: Duration) -> u64 {
```

Comment:
This is pretty much copied from https://github.com/mobilecoinfoundation/mobilecoin/blob/368d2e43879a7542f713bb8e16bf35255812b274/watcher/src/watcher_db.rs#L453
I think it's simple enough that this is acceptable.